### PR TITLE
feat: add support for JSONPath testcases

### PR DESCRIPTION
**Pull request by gchiesa**
## 🎯 What
Provide a concise description of the changes:
* ❓ What changes have you made? (High-level overview)
Implemented support for multiple TestCase analyzing the plan with JsonPath queryes

* 🎉 What does it mean to the user? (In plain English)

User can run some tests with the following model:

```go
//[...]
tc := []scenario.JsonPathTestCases{
		{
			TestName:           "isIRSA",
			JsonPathToCompare:  "{$.planned_values.outputs.is_irsa.value}",
			ExpectedValue:      false,
			TestType:           scenario.ShouldBeEqual,
			AllowDifferentType: false,
		},
		{
			TestName:           "isPodIdentity",
			JsonPathToCompare:  "{$.planned_values.outputs.is_pod_identity.value}",
			ExpectedValue:      true,
			TestType:           scenario.ShouldBeEqual,
			AllowDifferentType: false,
		},
}
s.Stg.PlanAndAssertJsonWithJsonPath(t, s.GetTerraformOptions(), tc)
```

### CODE REVIEW

File: pkg/scenario/stage.go [hunk 1]
```diff
@@ -166,6 +185,50 @@ func (c *StageClient) PlanWithSpecificVariableValueToExpect(t *testing.T, option
 	compareValues(t, actualValue, expectedValue, variable)
 }
 
+// PlanAndAssertJSONWithJSONPath performs JSON path planning and assertion in Go testing.
+//
+// t *testing.T: Testing object
+// options *terraform.Options: Terraform options
+// testCases []JSONPathTestCases: Array of JSON path test cases
+func (c *StageClient) PlanAndAssertJSONWithJSONPath(t *testing.T, options *terraform.Options, testCases []JSONPathTestCases) {
```
Comment (Excoriate) [inline pkg/scenario/stage.go:194]: This new function should be part of the `Stage` interface. By looking at your changes, it's not. Please, add it.
```go
type Stage interface {
	PlanAndAssertJSONWithJSONPath(t *testing.T, options *terraform.Options, testCases []JSONPathTestCases) 
}
```

File: pkg/scenario/stage.go [hunk 2]
```diff
@@ -166,6 +185,50 @@ func (c *StageClient) PlanWithSpecificVariableValueToExpect(t *testing.T, option
...
+	jsonPlan := terraform.InitAndPlanAndShow(t, options)
+
+	for _, testCase := range testCases {
+		t.Run(testCase.TestName, func(t *testing.T) {
+			var result []interface{}
+			k8s.UnmarshalJSONPath(t, []byte(jsonPlan), testCase.JSONPathToCompare, &result)
```
Comment (Excoriate) [inline pkg/scenario/stage.go:200]: Just curious, is there any specific reason why this library and this particular function from it is required instead of a 'vanilla' Unmarshal of a valid JSON into a go struct/object?

File: pkg/scenario/stage.go [hunk 3]
```diff
@@ -166,6 +185,50 @@ func (c *StageClient) PlanWithSpecificVariableValueToExpect(t *testing.T, option
...
+				if !testCase.AllowDifferentType {
+					assert.Equal(t, reflect.TypeOf(testCase.ExpectedValue).Kind(), reflect.TypeOf(v).Kind())
+				}
+				applyTestType(t, testCase.TestType, v, testCase.ExpectedValue, fmt.Sprintf("JSONPATH query: %s", testCase.JSONPathToCompare))
+			}
+		})
+	}
+}
+
+func applyTestType(t *testing.T, testType TestType, actual, expected interface{}, additionalMessage string) {
```
Comment (Excoriate) [inline pkg/scenario/stage.go:219]: Nice. It seems that this function can be used and even exposed as a package API, since it's quite generic. What do you think if you make it public? — Currently, it's private o this particular package.

File: pkg/scenario/stage.go [hunk 4]
```diff
@@ -166,6 +185,50 @@ func (c *StageClient) PlanWithSpecificVariableValueToExpect(t *testing.T, option
...
+	case ShouldNOTContain:
+		msg := fmt.Sprintf("Output is expected NOT to contains the value. Expected: %v, Actual: %v, %s", expected, actual, additionalMessage)
+		assert.NotContains(t, actual, expected, msg)
+	case ShouldBeEqual:
+		msg := fmt.Sprintf("Output did not match with expected value. Expected: %v, Actual: %v, %s", expected, actual, additionalMessage)
+		assert.EqualValuesf(t, actual, expected, msg)
+	}
+}
+
 func compareValues(t *testing.T, actual interface{}, expected, variableName string) {
```
Comment (Excoriate) [inline pkg/scenario/stage.go:233]: Curious, why require and not assert? I've noticed that in the applyTestType you're proposing assert, yet here require. If require doesn't pass, it's going to stop the whole test execution - is what you're looking to accomplish?

**Review by Excoriate — CHANGES_REQUESTED** [review]
Thanks so much for this nice contribution. It's indeed a great addition to the library! I left small comments for your revision.

File: pkg/scenario/stage.go (hunk 2 above)
Comment (gchiesa) [reply pkg/scenario/stage.go:200]: I just largerly used it in Terratest since it's a common way to query with JSONPath the plans. The package is still terratest so no big increase in dependencies, and it's maintained/extended upstream.

File: pkg/scenario/stage.go (hunk 4 above)
Comment (gchiesa) [reply pkg/scenario/stage.go:233]: I'm using assert instead require for that specific reaosn. I would like to have a global summary with what passed and what not, so `assert` seems the proper way to do it. But perhaps I'm not getting the point of your comment.

File: pkg/scenario/stage.go (hunk 3 above)
Comment (gchiesa) [reply pkg/scenario/stage.go:219]: we can have a next feature to eternalized it, however is in the scope of this JSONPath Test Case functionality so it's not meant to be public. Yet.

Comment (Excoriate) [reply pkg/scenario/stage.go:219]: That makes sense 🍺

**Review by Excoriate — APPROVED** [review]
